### PR TITLE
f_scores & sparse_f_scores

### PR DESCRIPTION
**Pull request by markub3327**
Hello,

@bhack @LukeWood 

   this is solution of #393. Please check it and I'll make a test for SparseFBetaScore.

Thanks.

### CODE REVIEW

File: keras_cv/metrics/f_scores.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+import tensorflow as tf
+import tensorflow.keras as keras
+
+
+class FBetaScore(keras.metrics.Metric):
+    r"""Computes F-Beta score.
+
+    It is the weighted harmonic mean of precision
+    and recall. Output range is `[0, 1]`. Works for
+    both multi-class and multi-label classification.
```
Comment (innat) [inline keras_cv/metrics/f_scores.py:24]: @markub3327 cc. @LukeWood 
Is this class compatible with object detection and segmentation task as well?

**Comment by bhack** [comment]
What is the status of this? We have another request on TF https://github.com/tensorflow/tensorflow/issues/56358

**Comment by LukeWood** [comment]
> What is the status of this? We are closing another request on TF [tensorflow/tensorflow#56358](https://github.com/tensorflow/tensorflow/issues/56358)

I've been prioritizing other issues over this for a bit, also deciding how components that support multiple label types: 

i.e. classification, object detection, and image segmentation should be structured

File: keras_cv/metrics/f_scores.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import tensorflow.keras as keras
+
+
+class FBetaScore(keras.metrics.Metric):
```
Comment (LukeWood) [inline keras_cv/metrics/f_scores.py:19]: Is there anything computer vision specific about this?  Could this also be used in NLP classification?  Perhaps this is better as a core Keras feature?

File: keras_cv/metrics/f_scores.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    $$
+    F_{\beta} = (1 + \beta^2) * \frac{\textrm{precision} * \textrm{recall}}{(\beta^2 \cdot \textrm{precision}) + \textrm{recall}}
+    $$
+    
+    Args:
+        num_classes: Number of unique classes in the dataset.
+        average: Type of averaging to be performed on data.
+            Acceptable values are `None`, `micro`, `macro` and
+            `weighted`. Default value is None.
+        beta: Determines the weight of precision and recall
```
Comment (LukeWood) [inline keras_cv/metrics/f_scores.py:35]: how does this parameter work?  Are numbers > 1 giving more weights to recall?

File: keras_cv/metrics/f_scores.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            in harmonic mean. Determines the weight given to the
+            precision and recall. Default value is 1.
+        threshold: Elements of `y_pred` greater than threshold are
```
Comment (LukeWood) [inline keras_cv/metrics/f_scores.py:38]: i think we call this "from_logits" in keras

File: keras_cv/metrics/f_scores.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+            converted to be 1, and the rest 0. If threshold is
+            None, the argmax is converted to 1, and the rest 0.
+        name: (Optional) String name of the metric instance.
```
Comment (LukeWood) [inline keras_cv/metrics/f_scores.py:41]: You can just support name and dtype from **kwargs

File: keras_cv/metrics/f_scores.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            "average": self.average,
+            "beta": self.beta,
+            "threshold": self.threshold,
+        }
+
+        base_config = super().get_config()
+        return {**base_config, **config}
+
+
+class F1Score(FBetaScore):
```
Comment (LukeWood) [inline keras_cv/metrics/f_scores.py:207]: lets use one file per metric.  it is easier to stay organized and read/review code in that way.  feel free to make an `f_scores` subdirectory

File: keras_cv/metrics/f_scores.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/metrics/f_scores.py:24]: These would probably need a separate implementation I'd think

**Review by LukeWood — CHANGES_REQUESTED** [review]

File: keras_cv/metrics/f_scores.py (hunk 2 above)
Comment (bhack) [reply keras_cv/metrics/f_scores.py:19]: What Is 2022 opinion of the team?

https://github.com/keras-team/keras/issues/5794

Comment (LukeWood) [reply keras_cv/metrics/f_scores.py:19]: Looks like Precision was reintroduced: https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Precision

(wasn’t involved in anything regard this metric so perhaps I am incorrect)

File: keras_cv/metrics/f_scores.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/metrics/f_scores.py:38]: Nevermind, I was incorrect.

File: keras_cv/metrics/f_scores.py (hunk 3 above)
Comment (markub3327) [reply keras_cv/metrics/f_scores.py:35]: I think yes.

File: keras_cv/metrics/f_scores.py (hunk 5 above)
Comment (markub3327) [reply keras_cv/metrics/f_scores.py:41]: Ok.

**Comment by innat** [comment]
Why is it required to have `f_1/2_score` if we can get them with general `f_beta_score`?

File: keras_cv/metrics/f_scores.py (hunk 2 above)
Comment (bhack) [reply keras_cv/metrics/f_scores.py]: Also recall but you need to check internally with you team.

**Comment by markub3327** [comment]
@innat 

It's easier for the newbie users to understand the CV.

**Comment by bhack** [comment]
> Why is it required to have f_1/2_score if we can get them with a general f_beta_score? It doesn't seem to need various beta=n versions and thus will reduce implemented cost and clean API.

It could be similar to passing single RGB values or Colormapping: https://github.com/keras-team/keras-cv/pull/433

**Comment by markub3327** [comment]
Hi @bhack @LukeWood, all tests are passed. You can merge it. Thanks!

**Comment by bhack** [comment]
Do we have a review here?

File: keras_cv/metrics/f_scores.py (hunk 2 above)
Comment (tanzhenyu) [reply keras_cv/metrics/f_scores.py]: Both Precision and Recall are in Keras now. We should add F Score

**Comment by innat** [comment]
@markub3327 
Could you please test your implementation with this https://github.com/keras-team/keras/issues/17603 ? (As this implementation is quite similar to tfa.f1 score.)

**Comment by markub3327** [comment]
@innat 
I have available PC with only one GPU.

**Comment by innat** [comment]
You can still try with mirror or single strategy scope. Or, use kaggle env, it provides multi gpu for free and tf 2.11.

**Comment by markub3327** [comment]
@innat @bhack 

All is working for me. Here is notebook: https://www.kaggle.com/code/martinkubovksk/notebooke5641a9136

Here is the result:
> Number of devices: 2
>
> Epoch 1/5
> 2/2 [==============================] - 15s 608ms/step - loss: 0.8687 - f1_score: 0.6667 - val_loss: 0.9360 - val_f1_score: 0.6667
> Epoch 2/5
> 2/2 [==============================] - 0s 177ms/step - loss: 0.8406 - f1_score: 0.6667 - val_loss: 0.9011 - val_f1_score: 0.6667
> Epoch 3/5
> 2/2 [==============================] - 0s 173ms/step - loss: 0.8150 - f1_score: 0.6667 - val_loss: 0.8684 - val_f1_score: 0.6667
> Epoch 4/5
> 2/2 [==============================] - 0s 173ms/step - loss: 0.7930 - f1_score: 0.6667 - val_loss: 0.8386 - val_f1_score: 0.6667
> Epoch 5/5
> 2/2 [==============================] - 0s 165ms/step - loss: 0.7698 - f1_score: 0.6667 - val_loss: 0.8110 - val_f1_score: 0.6667
> [8]:
> <keras.callbacks.History at 0x7fe44859cfd0>
